source/mir/sparse/package.d

```diff
@@ -0,0 +1,614 @@
+    static assert(isRandomAccessRange!(Sparse!(2, double)));
+
+    import mir.ndslice.slice: Slice, DeepElementType;
+    static assert(is(Sparse!(2, double) : Slice!(2, Range), Range));
+    static assert(is(DeepElementType!(Sparse!(2, double)) == double));
+}
+
+/++
+Sparse Slice in Dictionary of Keys (DOK) format.
++/
+alias Sparse(size_t N, T) = Slice!(N, SparseMap!T);
+
+/++
+SparseMap is a range, which is used internally by $(LREF Sparse).
++/
+struct SparseMap(T)
+{
+    ///
+    T[size_t] table;
+
+    ///
+    auto save() @property
+    {
+        return this;
+    }
+
+    ///
+    T opIndex(size_t index)
+    {
+        static if (isScalarType!T)
+            return table.get(index, cast(T)0);
+        else
+            return table.get(index, null);
+    }
+
+    ///
+    T opIndexAssign(T value, size_t index)
+    {
+        static if (isScalarType!T)
+        {
+            if (value != 0)
+            {
+                table[index] = value;
+            }
+        }
+        else
+        {
+            if (value !is null)
+            {
+                table[index] = value;
+            }
+        }
+        return value;
+    }
+
+    ///
+    T opIndexUnary(string op)(size_t index)
+        if (op == `++` || op == `--`)
+    {
+        mixin (`auto value = ` ~ op ~ `table[index];`);
+        static if (isScalarType!T)
+        {
+            if (value == 0)
+            {
+                table.remove(index);
+            }
+        }
+        else
+        {
+            if (value is null)
+            {
+                table.remove(index);
+            }
+        }
+        return value;
+    }
+
+    ///
+    T opIndexOpAssign(string op)(T value, size_t index)
+        if (op == `+` || op == `-`)
+    {
+        mixin (`value = table[index] ` ~ op ~ `= value;`); // this works
+        static if (isScalarType!T)
+        {
+            if (value == 0)
+            {
+                table.remove(index);
+            }
+        }
+        else
+        {
+            if (value is null)
+            {
+                table.remove(index);
+            }
+        }
+        return value;
+    }
+}
+
+/++
++/
+struct CoordinateValue(size_t N, T)
+{
+    ///
+    size_t[N] index;
+    
+    ///
+    T value;
+
+    ///
+    sizediff_t opCmp()(auto ref const typeof(this) rht) const
+    {
+        return cmpCoo(this.index, rht.index);
+    }
+}
+
+private sizediff_t cmpCoo(size_t N)(const auto ref size_t[N] a, const auto ref size_t[N] b)
+{
+    foreach(i; Iota!(0, N))
+        if(auto d = a[i] - b[i])
+            return d;
+    return 0;
+}
+
+/++
++/
+auto byCoordinateValue(S : Slice!(N, R), size_t N, R : SparseMap!T, T)(S slice)
+{
+    static struct CoordinateValues
+    {
+        static if(N > 1)
+            private sizediff_t[N-1] _strides;
+        mixin _sparse_range_methods!(N, T);
+        private typeof(Sparse!(N, T).init.ptr.range.table.byKeyValue()) _range;
+
+        auto front() @property
```

Comment:
If this returns a class than class would be const and user would not be able to change it